velox/functions/sparksql/ConcatWs.cpp

```diff
@@ -0,0 +1,391 @@
+  }
+
+  // Calculate the total number of bytes in the result.
+  size_t calculateTotalResultBytes(
+      const SelectivityVector& rows,
+      exec::EvalCtx& context,
+      std::vector<exec::LocalDecodedVector>& decodedArrays,
+      const std::vector<std::optional<std::string>>& constantStrings,
+      const std::vector<exec::LocalDecodedVector>& decodedStringArgs,
+      const exec::LocalDecodedVector& decodedSeparator) const {
+    auto arrayArgNum = decodedArrays.size();
+    std::vector<const ArrayVector*> arrayVectors;
+    std::vector<DecodedVector> elementsDecodedVectors;
+    for (auto i = 0; i < arrayArgNum; ++i) {
+      auto arrayVector = decodedArrays[i].get()->base()->as<ArrayVector>();
+      arrayVectors.push_back(arrayVector);
+      SelectivityVector nestedRows(arrayVector->elements()->size());
+      elementsDecodedVectors.emplace_back(*arrayVector->elements(), nestedRows);
+    }
+
+    size_t totalResultBytes = 0;
+    rows.applyToSelected([&](auto row) {
+      // NULL separator produces NULL result.
+      if (!isConstantSeparator() && decodedSeparator->isNullAt(row)) {
+        return;
+      }
+      int32_t allElements = 0;
+      // Calculate size for array columns data.
+      for (int i = 0; i < arrayArgNum; i++) {
+        if (decodedArrays[i]->isNullAt(row)) {
+          continue;
+        }
+        auto indices = decodedArrays[i].get()->indices();
+        auto size = arrayVectors[i]->sizeAt(indices[row]);
+        auto offset = arrayVectors[i]->offsetAt(indices[row]);
+
+        for (int j = 0; j < size; ++j) {
+          if (!elementsDecodedVectors[i].isNullAt(offset + j)) {
+            auto element =
+                elementsDecodedVectors[i].valueAt<StringView>(offset + j);
+            // No matter empty string or not.
+            ++allElements;
+            totalResultBytes += element.size();
+          }
+        }
+      }
+
+      // Calculate size for string arg.
+      auto it = decodedStringArgs.begin();
+      for (int i = 0; i < constantStrings.size(); i++) {
+        StringView value;
+        if (constantStrings[i].has_value()) {
+          value = StringView(*constantStrings[i]);
+        } else {
+          VELOX_CHECK(
+              it < decodedStringArgs.end(),
+              "Unexpected end when iterating over decodedStringArgs.");
+          // Skip NULL.
+          if ((*it)->isNullAt(row)) {
+            ++it;
+            continue;
+          }
+          value = (*it++)->valueAt<StringView>(row);
+        }
+        // No matter empty string or not.
+        allElements++;
+        totalResultBytes += value.size();
+      }
+
+      auto separatorSize = isConstantSeparator()
+          ? separator_.value().size()
+          : decodedSeparator->valueAt<StringView>(row).size();
+
+      if (allElements > 1 && separatorSize > 0) {
+        totalResultBytes += (allElements - 1) * separatorSize;
+      }
+    });
+    return totalResultBytes;
+  }
+
+  void initVectors(
```

Comment:
Perhaps add some comments for the optimization of concating consecutive constant string args in advance.